### ISSUE TITLE
register addressing

### ISSUE THREAD

**Issue by rob040** [CMB27/ModbusRTUSlave#18]
From the code and sample, it appears that all uses of register addresses start at zero and end at { numCoils, numDiscreteInputs, numHoldingRegisters, numInputRegisters }
This is not common practice on Modbus.
How did you envision this?

**Comment by CMB27** [comment]
That is correct. I am aware that this is not especially common practice, but to my knowledge, it is allowed by the Modbus specification. In fact there is a diagram on page 8 of `MODBUS Application Protocol Specification V1.1b3` showing an example where the range of addresses for each of the different data types (coils, registers, etc) all start with 1. It is also common practice in programming for lists and arrays to start with 0.

There is indeed checking of valid/invalid data addresses. The check is very simple. Since the `address` data type is an unsigned integer (no negative values) and starts at 0, all that needs to be done is to check that the data address value does not exceed the number of addresses (numCoils, numDiscreteInputs, etc.) set up for its respective data type (coils, discrete inputs, etc.).
